server/schedule/filter/filters.go

```diff
@@ -601,3 +601,77 @@ const (
 
 var allSpecialUses = []string{SpecialUseHotRegion, SpecialUseReserved}
 var allSpeicalEngines = []string{EngineTiFlash}
+
+type isolationFilter struct {
+	scope          string
+	locationLabels []string
+	constraintSet  [][]string
+}
+
+// NewIsolationFilter creates a filter that filters out stores with isolationLevel
+// For example, a region has 3 replicas in z1, z2 and z3 individually.
+// With isolationLevel = zone, if the region on z1 is down, we need to filter out z2 and z3
+// because these two zones already have one of the region's replicas on them.
+// We need to choose a store on z1 or z4 to place the new replica to meet the isolationLevel explicitly and forcibly.
+func NewIsolationFilter(scope, isolationLevel string, locationLabels []string, regionStores []*core.StoreInfo) Filter {
+	isolationFilter := &isolationFilter{
+		scope:          scope,
+		locationLabels: locationLabels,
+		constraintSet:  make([][]string, 0),
+	}
+	if len(isolationLevel) <= 0 || len(locationLabels) <= 0 {
+		return isolationFilter
+	}
+	// Get which idx this isolationLevel at according to locationLabels
+	var isolationLevelIdx int
+	for level, label := range locationLabels {
+		if label == isolationLevel {
+			isolationLevelIdx = level
+			break
+		}
+	}
+	// Collect all constraints for given isolationLevel
+	for _, regionStore := range regionStores {
+		constraintList := make([]string, 0)
+		for i := 0; i <= isolationLevelIdx; i++ {
+			constraintList = append(constraintList, regionStore.GetLabelValue(locationLabels[i]))
+		}
+		isolationFilter.constraintSet = append(isolationFilter.constraintSet, constraintList)
+	}
+	return isolationFilter
+}
+
+func (f *isolationFilter) Scope() string {
+	return f.scope
+}
+
+func (f *isolationFilter) Type() string {
+	return "isolation-filter"
+}
+
+func (f *isolationFilter) Source(opt opt.Options, store *core.StoreInfo) bool {
+	return true
+}
+
+func (f *isolationFilter) Target(opt opt.Options, store *core.StoreInfo) bool {
+	// No isolation constraint to fit
+	if len(f.constraintSet) <= 0 {
+		return true
+	}
+	for _, constrainList := range f.constraintSet {
+		match := true
+		for idx, constraint := range constrainList {
+			// Check every constraint in constrainList
+			labelConstraint := &placement.LabelConstraint{
```

Comment:
I meant that you can compare label values directly without introducing `LabelConstraint` here.

Reply:
You mean just use `GetLabelValue` to check label and value manually?

Reply:
I got it. It's indeed much simpler than using `LabelConstraint` in this case...